hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestMaxResultsPerColumnFamily.java

```diff
@@ -0,0 +1,300 @@
+import org.apache.hadoop.hbase.TableName;
+import org.apache.hadoop.hbase.client.Admin;
+import org.apache.hadoop.hbase.client.ColumnFamilyDescriptorBuilder;
+import org.apache.hadoop.hbase.client.Durability;
+import org.apache.hadoop.hbase.client.Put;
+import org.apache.hadoop.hbase.client.Result;
+import org.apache.hadoop.hbase.client.ResultScanner;
+import org.apache.hadoop.hbase.client.Scan;
+import org.apache.hadoop.hbase.client.Table;
+import org.apache.hadoop.hbase.client.TableDescriptor;
+import org.apache.hadoop.hbase.client.TableDescriptorBuilder;
+import org.apache.hadoop.hbase.testclassification.MediumTests;
+import org.apache.hadoop.hbase.testclassification.RestTests;
+import org.apache.hadoop.hbase.util.Bytes;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@Category({RestTests.class, MediumTests.class})
+public class TestMaxResultsPerColumnFamily {
+
+  @ClassRule
+  public static final HBaseClassTestRule CLASS_RULE =
+    HBaseClassTestRule.forClass(TestMaxResultsPerColumnFamily.class);
+
+  private static final Logger LOG = LoggerFactory.getLogger(
+    TestMaxResultsPerColumnFamily.class);
+
+  private static final byte [][] FAMILIES = {
+    Bytes.toBytes("1"), Bytes.toBytes("2")
+  };
+
+  private static final byte [][] VALUES = {
+    Bytes.toBytes("testValueOne"), Bytes.toBytes("testValueTwo")
+  };
+
+  private static final HBaseTestingUtil hbaseTestingUtility = new HBaseTestingUtil();
+
+  public static byte[] toByteArray(String s) {
+    return Bytes.toBytes(s);
+  }
+
+  @BeforeClass
+  public static void setUpBeforeClass() throws Exception {
+    hbaseTestingUtility.startMiniCluster(3);
+  }
+
+  @AfterClass
+  public static void tearDownAfterClass() throws Exception {
+    hbaseTestingUtility.shutdownMiniCluster();
+  }
+
+  @Test
+  public void tesSetMaxResultsPerColumnFamily() throws Exception {
+    Admin admin = hbaseTestingUtility.getAdmin();
+    TableName tableName = TableName.valueOf("TestScannersWithFilters1");
+    if (!admin.tableExists(tableName)) {
+      ColumnFamilyDescriptorBuilder cfBuilder0 =
+        ColumnFamilyDescriptorBuilder.newBuilder(FAMILIES[0]);
+      ColumnFamilyDescriptorBuilder cfBuilder1 =
+        ColumnFamilyDescriptorBuilder.newBuilder(FAMILIES[1]);
+
+      TableDescriptor tableDescriptor =
+        TableDescriptorBuilder.newBuilder(tableName).setColumnFamily(cfBuilder0.build())
+          .setColumnFamily(cfBuilder1.build()).build();
+      admin.createTable(tableDescriptor);
+      Table table = hbaseTestingUtility.getConnection().getTable(tableName);
+
+      for(int i = 0; i<20000; i++) {
+        byte [] ROW = toByteArray("" + i);
+        Put p = new Put(ROW);
+        p.setDurability(Durability.SKIP_WAL);
+
+        int rows = 4;
+        if (i % 12 == 5 || i % 17 == 3) {
```

Comment:
Do we really need these magics to reproduce the bug?

Reply:
We don't need that.
I just wanted to create data where the number of columns are not the same for every row.
Thought it makes the data more realistic and it helped me to reproduce the issue in a test.
Note that testSetMaxResultsPerColumnFamilySimpleBig also reproduced the issue without that magic.